### PR TITLE
Refactor rulerstore packages

### PR DESCRIPTION
**Pull request by pracucci**
**What this PR does**:
Similarly to the package refactoring done in #3881, in this PR I'm proposing the same refactoring (same structure) for the rulestore. This should help to better reorganize packages.

The diff is huge and I'm sorry for that, but most changes have been done with the Goland IDE refactoring tool and splitted in different commits:
- Moved bucketclient implementation to pkg/ruler/rulestore/bucketclient
- Moved local and objectclient rule store under pkg/ruler/rulestore/...
- Renamed NewRuleStorage() to NewLegacyRuleStore() 
- Renamed pkg/ruler/rules to pkg/ruler/rulespb

**Which issue(s) this PR fixes**:
N/A

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/ruler/rulestore/config.go [hunk 1]
```diff
@@ -31,28 +26,3 @@ func (cfg *Config) RegisterFlags(f *flag.FlagSet) {
 	cfg.ConfigDB.RegisterFlagsWithPrefix(prefix, f)
 	cfg.RegisterFlagsWithPrefix(prefix, f)
 }
-
-// NewRuleStore creates a new bucket client based on the configured backend
-func NewRuleStore(ctx context.Context, cfg Config, cfgProvider bucket.TenantConfigProvider, logger log.Logger, reg prometheus.Registerer) (rules.RuleStore, error) {
```
Comment (pracucci) [inline pkg/ruler/rulestore/config.go:36]: Moved to `pkg/ruler/storage.go` to avoid circular dependencies.

**Review by gotjosh — APPROVED** [review]
YES! (Assuming test and lint pass)

File: pkg/ruler/ruler.go [hunk 2]
```diff
@@ -28,8 +28,9 @@ import (
 	"github.com/cortexproject/cortex/pkg/ingester/client"
 	"github.com/cortexproject/cortex/pkg/ring"
 	"github.com/cortexproject/cortex/pkg/ring/kv"
-	"github.com/cortexproject/cortex/pkg/ruler/rules"
-	store "github.com/cortexproject/cortex/pkg/ruler/rules"
+	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
+	store "github.com/cortexproject/cortex/pkg/ruler/rulespb"
```
Comment (jtlisi) [inline pkg/ruler/ruler.go:32]: **suggestion:** we may be able to trim this down

```suggestion
	"github.com/cortexproject/cortex/pkg/ruler/rulespb"
```

**Review by jtlisi — APPROVED** [review]
Awesome! LGTM besides an optional suggestion

**Review by pstibrany — APPROVED** [review]
LGTM, thank you.

File: pkg/ruler/ruler.go (hunk 2 above)
Comment (pracucci) [reply pkg/ruler/ruler.go]: Definitely! I didn't notice it, thanks! (and done)

**Pull request merged**